server/paymentProviders/thegivingblock/index.js

```diff
@@ -7,27 +7,53 @@ import fetch from 'node-fetch';
 import orderStatus from '../../constants/order_status';
 import { TransactionTypes } from '../../constants/transactions';
 import { getFxRate } from '../../lib/currency';
+import logger from '../../lib/logger';
 import { getHostFee, getHostFeeSharePercent } from '../../lib/payments';
 import models from '../../models';
 
 const AES_ENCRYPTION_KEY = config.thegivingblock.aesEncryptionKey;
 const AES_ENCRYPTION_IV = config.thegivingblock.aesEncryptionIv;
 const AES_ENCRYPTION_METHOD = config.thegivingblock.aesEncryptionMethod;
 const API_URL = config.thegivingblock.apiUrl;
+const GIVINGBLOCK_USERNAME = config.thegivingblock.username;
+const GIVINGBLOCK_PASSWORD = config.thegivingblock.password;
+const MAX_REQUEST_RETRIES = 3;
 
-async function apiRequest(path, options = {}) {
+async function apiRequest(path, options = {}, account) {
   const response = await fetch(`${API_URL}${path}`, options);
   const result = await response.json();
 
-  if (result.data.errorMessage) {
-    throw new Error(`The Giving Block: ${result.data.errorMessage}`);
-  }
-
+  await handleErrors(result, path, options, account);
   // console.log(result);
 
   return result.data;
 }
 
+/*
+ * Whenever and api request is made we check if access token is expired and if so we login again.
+ * Refer: https://app.gitbook.com/@the-giving-block/s/public-api-documentation/#authentication-flow.
+ * Access tokens are only valid for 2 hours.
+ */
+async function handleErrors(result, path, options = {}, account = null) {
+  if (result.data.errorMessage) {
+    if (result.data.meta.errorCode === 'INVALID_JWT_TOKEN' && account) {
+      logger.debug('Access token is invalid. Requesting a new one.');
+      let error;
+      for (let retries = 0; retries < MAX_REQUEST_RETRIES; retries++) {
+        try {
+          const { accessToken, refreshToken } = await login(GIVINGBLOCK_USERNAME, GIVINGBLOCK_PASSWORD);
+          await account.update({ data: { ...account.data, accessToken, refreshToken } });
+          return await apiRequest(path, options);
```

Comment:
You're still not updating the accessToken in the apiRequest, how do you expect this to work?

Reply:
@znarf : I am was trying some things here. This isn't done yet; should have marked this PR as draft earlier. I'll ping you once completed. I am trying to find a way to update the accessToken as well as resend the request. It's a bit tricky as in some requests we send the refresh token whereas in the deposit request we send the access token.